semantics/src/traversal/expressions.rs

```diff
@@ -256,6 +256,31 @@ fn expr_bin_operation(
     unreachable!()
 }
 
+fn expr_call(
+    scope: Shared<FunctionScope>,
+    context: Shared<Context>,
+    exp: &Spanned<fe::Expr>,
+) -> Result<ExpressionAttributes, SemanticError> {
+    if let fe::Expr::Call { func, args } = &exp.node {
+
+        if let fe::Expr::Attribute { value:_, attr } = &func.node {
+            let contract_scope = &scope.borrow().parent;
+            let called_func = contract_scope.borrow().def(attr.node.to_string());
+            panic!("{:?}, {:?}, {}", contract_scope.borrow().defs, called_func, attr.node.to_string());
```

Comment:
@g-r-a-n-t If I understood you correctly, then given the following file:

```
contract Foo:
    pub def bar() -> u256:
        return self.foo()

    pub def foo() -> u256:
        return 42
``` 

We should be performing a check on the definition of `foo` here to throw an error in case it turns out it doesn't have an return value. Because in that case the expression `return self.foo()` wouldn't be legal.

I'm trying to do that so I'm going up to the contract_scope with the intention to lookup the definition of `foo`. However, to my surprise it doesn't exist here.

The result of the panic is:

>`thread 'test_method_return::case_01' panicked at '{"bar": Function { params: [], returns: Some(Base(U256)) }}, None, foo', semantics/src/traversal/expressions.rs:269:13
`

So, there is a definition for `bar` but not for `foo`.

Am I doing it wrong?

Reply:
At the moment, function definitions are [analyzed one at a time](https://github.com/ethereum/fe/blob/de034e05af021904be516ebd93cf9ed63640b1d5/semantics/src/traversal/contracts.rs#L39) from top to bottom. So, since `foo` is defined after `bar`, we can't see `foo` in `bar`s function body.

We should modify the [contract analysis](https://github.com/ethereum/fe/blob/de034e05af021904be516ebd93cf9ed63640b1d5/semantics/src/traversal/contracts.rs#L29) function to instead do an initial sweep on all function definitions that adds their signatures to the scope.

If you'd like, we can just swap the order of the functions in this contract and create an issue for function signature analysis.